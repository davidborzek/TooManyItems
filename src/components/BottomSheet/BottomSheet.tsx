import {
  Modal,
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native';

import React from 'react';

export type BottomSheetItem = {
  text: string;
  disabled?: boolean;
  color?: string;
  onPress: () => void;
};

type Props = {
  items: BottomSheetItem[];
  visible?: boolean;
  onClose?: () => void;
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'rgba(0,0,0,0.5)',
    flex: 1,
    justifyContent: 'flex-end',
  },
  overlay: { flex: 1 },
  sheet: {
    backgroundColor: 'white',
  },
});

/**
 * Ein wiederverwendbarer Component welcher mehrere AuswahlmÃ¶glichkeiten am unteren Bildschirmrand bereitstellt.
 */
export default function BottomSheet({ items, visible, onClose }: Props) {
  const handleItemClick = (item: BottomSheetItem) => {
    if (onClose) {
      onClose();
    }
    item.onPress();
  };

  const renderItems = () => {
    return items.map((item) => (
      <TouchableOpacity
        key={item.text}
        onPress={() => handleItemClick(item)}
        disabled={item.disabled}
      >
        <Text
          style={{
            color: item.color,
            opacity: item.disabled ? 0.3 : 1,
            padding: 16,
          }}
        >
          {item.text}
        </Text>
      </TouchableOpacity>
    ));
  };

  return (
    <Modal
      animationType="fade"
      visible={visible}
      onRequestClose={onClose}
      transparent
      statusBarTranslucent
    >
      <View style={styles.container}>
        <TouchableWithoutFeedback onPress={onClose}>
          <View style={styles.overlay} />
        </TouchableWithoutFeedback>
        <View style={styles.sheet}>{renderItems()}</View>
      </View>
    </Modal>
  );
}
