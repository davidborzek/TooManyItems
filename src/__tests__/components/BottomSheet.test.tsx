import BottomSheet, {
  BottomSheetItem,
} from '../../components/BottomSheet/BottomSheet';

import React from 'react';
import renderer from 'react-test-renderer';

const items: BottomSheetItem[] = [
  {
    onPress: () => console.log('do something'),
    text: 'Some text',
  },
  {
    onPress: () => console.log('do something'),
    disabled: true,
    text: 'Some disabled text',
  },
  {
    color: 'red',
    onPress: () => console.log('do something'),
    text: 'Some red text',
  },
  {
    color: 'red',
    disabled: true,
    onPress: () => console.log('do something'),
    text: 'Some red disabled text',
  },
];

it('BottomSheet renders correctly and is closed', () => {
  const tree = renderer.create(<BottomSheet items={items} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('BottomSheet renders correctly and is opened', () => {
  const tree = renderer.create(<BottomSheet items={items} visible />).toJSON();
  expect(tree).toMatchSnapshot();
});
