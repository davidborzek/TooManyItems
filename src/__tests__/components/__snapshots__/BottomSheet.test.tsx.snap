// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomSheet renders correctly and is closed 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  statusBarTranslucent={true}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0.5)",
        "flex": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "white",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": undefined,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          Some text
        </Text>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": undefined,
              "opacity": 0.3,
              "padding": 16,
            }
          }
        >
          Some disabled text
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "red",
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          Some red text
        </Text>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "red",
              "opacity": 0.3,
              "padding": 16,
            }
          }
        >
          Some red disabled text
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;

exports[`BottomSheet renders correctly and is opened 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  statusBarTranslucent={true}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0,0,0,0.5)",
        "flex": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <View
      style={
        Object {
          "backgroundColor": "white",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": undefined,
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          Some text
        </Text>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": undefined,
              "opacity": 0.3,
              "padding": 16,
            }
          }
        >
          Some disabled text
        </Text>
      </View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "red",
              "opacity": 1,
              "padding": 16,
            }
          }
        >
          Some red text
        </Text>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "red",
              "opacity": 0.3,
              "padding": 16,
            }
          }
        >
          Some red disabled text
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;
